### PR TITLE
[512] The great \ref purge

### PR DESCRIPTION
**Pull request by ninjanomnom**
This change is a result of the default \ref behaviour changing in 512. \ref no longer uses the object's tag var if available. REF() replicates that behavior and this also adds a new datum var for determining whether something should use its tag which defaults to FALSE.

Replaced:
`\\ref\[([^\[]+?)\]`
With:
`[REF($1)]`

Uses of \ref like `text("...\ref[]...", src)` Were manually fixed as there weren't many.

### CODE REVIEW

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -1396,3 +1396,16 @@ GLOBAL_DATUM_INIT(dview_mob, /mob/dview, new)
 //checks if a turf is in the planet z list.
 /proc/turf_z_is_planet(turf/T)
 	return GLOB.z_is_planet["[T.z]"]
+
+// \ref behaviour got changed in 512 so this is necesary to replicate old behaviour.
+// If it ever becomes necesary to get a more performant REF(), this lies here in wait
+// #define REF(thing) (thing && istype(thing, /datum) && thing:use_tag && thing:tag ? "[thing:tag]" : "\ref[thing]")
+
+proc/REF(input)
```
Comment (lzimann) [inline code/__HELPERS/unsorted.dm]: missing `/` at the start !

Comment (ninjanomnom) [reply code/__HELPERS/unsorted.dm]: woops

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -1396,3 +1396,16 @@ GLOBAL_DATUM_INIT(dview_mob, /mob/dview, new)
 //checks if a turf is in the planet z list.
 /proc/turf_z_is_planet(turf/T)
 	return GLOB.z_is_planet["[T.z]"]
+
+// \ref behaviour got changed in 512 so this is necesary to replicate old behaviour.
+// If it ever becomes necesary to get a more performant REF(), this lies here in wait
+// #define REF(thing) (thing && istype(thing, /datum) && thing:use_tag && thing:tag ? "[thing:tag]" : "\ref[thing]")
+
+/proc/REF(input)
+	if(istype(input, /datum))
```
Comment (Cyberboss) [inline code/__HELPERS/unsorted.dm]: better off using a typecache methinks

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/__HELPERS/unsorted.dm (hunk 2 above)
Comment (ninjanomnom) [reply code/__HELPERS/unsorted.dm]: Probably yeah, that's a massive typecache though.

Comment (MrStonedOne) [reply code/__HELPERS/unsorted.dm]: no, type caches only work if you can assume it is a datum, but \ref has to work properly on strings and numbers and lists for certain things

Comment (ninjanomnom) [reply code/__HELPERS/unsorted.dm]: I just took a shit and realized that but you ninja'd me >:(

File: code/__HELPERS/unsorted.dm [hunk 3]
```diff
@@ -1396,3 +1396,16 @@ GLOBAL_DATUM_INIT(dview_mob, /mob/dview, new)
...
+		var/datum/thing = input
+		if(thing && thing.use_tag)
```
Comment (MrStonedOne) [inline code/__HELPERS/unsorted.dm]: istype() returns false on nulls so the `if (thing)` is unneeded.

File: code/__HELPERS/unsorted.dm (hunk 2 above)
Comment (MoreRobustThanYou) [reply code/__HELPERS/unsorted.dm]: Ba dum tsh



Now I gotta shit too

**Comment by JJRcop** [comment]
What's holding this?

**Pull request merged**